solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java

```diff
@@ -0,0 +1,328 @@
+   * URL.
+   *
+   * <p>Take note that the ShardRequest can be reused/shared by sending to many different shards.
+   * Therefore, we need this class as a reference back to the corresponding ShardRequest when such
+   * future is completed.
+   */
+  class ShardRequestTrackingCallback implements HttpShardHandler.ShardRequestCallback {
+    private final ShardRequest shardRequest;
+    private final Future<LBSolrClient.Rsp> future;
+    private final List<String> shardUrls;
+
+    ShardRequestTrackingCallback(
+        Future<LBSolrClient.Rsp> future, ShardRequest shardRequest, List<String> shardUrls) {
+      this.future = future;
+      this.shardRequest = shardRequest;
+      this.shardUrls = shardUrls;
+    }
+
+    @Override
+    public void onResponse(LBSolrClient.Rsp response, long elapsedTime) {
+      onComplete(elapsedTime, response.getServer());
+    }
+
+    @Override
+    public void onException(Throwable exception, long elapsedTime) {
+      // TODO is it possible to infer the selected node? If it has timed out, perhaps we can assume
+      // all shardNodes are slow?
+      onComplete(elapsedTime, null);
+    }
+
+    private void onComplete(long elapsedTime, String selectedShardUrl) {
+      try {
+        actorsByShardRequest.compute(
+            shardRequest,
+            (k, actors) -> {
+              if (actors != null) {
+                actors.removeIf(
+                    actor ->
+                        actor.onRequestCompleted(selectedShardUrl, shardUrls, future, elapsedTime));
+                if (actors.isEmpty()) {
+                  return null;
+                }
+              }
+              return actors;
+            });
+      } catch (Exception e) {
+        log.warn("Failed to notify stats to slowNodeDetector", e);
+      }
+    }
+  }
+}
+
+/** This gets notified when such ShardRequest instance is submitted/completed to/on any shard url */
+interface ShardRequestActor {
+  void onRequestSubmitted(List<String> shardUrls, Future<?> future);
+
+  boolean onRequestCompleted(
+      String selectedShardUrl, List<String> shardUrls, Future<?> future, long timeElapsed);
+}
+
+class Util {
+  static String getNode(String urlString) {
+    try {
+      URL url = new URL(urlString);
+      return url.getAuthority();
+    } catch (MalformedURLException e) {
+      throw new IllegalArgumentException("Invalid URL: " + urlString, e);
+    }
+  }
+}
+
+/**
+ * Tied to a single ShardRequest instance, which can be retried/submitted to many shard urls.
+ *
+ * <p>This actor keeps the stats for each shard response latency and notify the SlowNodeDetector
+ * when all of them complete
+ */
+class NodeStatsCollector implements ShardRequestActor {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private final SlowNodeDetector detector;
+  private final AtomicInteger pendingCount = new AtomicInteger(0);
```

Comment:
Is there a reason that we track `pendingCount` at the `ShardRequestActor` level, as opposed to augmenting the ArrayList to carry `pendingCount` state? If that's possible, I think doing so would also address the ConcurrentModificationException concern on ArrayList. (I think there'd still be a race condition, but basically benign)

Reply:
Yep. Removed the pendingCount from NodeStatsCollector in the latest implementation, the higher level logic now determines whenever `close` should be invoked on the actor, hence flushing the data.